File: aizu/tree/ALDS1_7_A.rb
class Node

  attr_accessor:id, :parent, :child_left, :right_sibling, :depth

  def initialize(id)
    @id = id
    @parent = nil
    @child_left = nil
    @right_sibling = nil
    @depth = 0
  end

  def depth
    if @parent.nil?
      0
    else
      @parent.depth + 1
    end
  end

  def type
    if @parent == nil
      "root"
    elsif @child_left.nil?
      "leaf"
    else
      "internal node"
    end
  end

  def to_s
    parent_id =
      if @parent.nil?
        -1
      else
        @parent.id
      end
    "node #{@id}: parent = #{parent_id}, depth = #{depth}, #{type}, [#{children_to_s.join(', ')}]"
  end

  def children_to_s
    ss = []
    if @child_left.nil?
      ss
    else
      a = @child_left
      while a != nil
        ss.push(a.id.to_s)
        a = a.right_sibling
      end
      ss
    end
  end
end

n = STDIN.gets.to_i

tree = []
for i in 0..(n - 1)
  as = STDIN.gets.split(' ')
  id = as[0].to_i
  m = as[1].to_i
  children = as.drop(2).map { |x| x.to_i }

  node = tree[id]
  if node.nil?
    node = Node.new(id)
    tree[id] = node
  end

  l = 0
  children.each_with_index { |c, j|
    child_node = tree[c]
    if child_node.nil?
      child_node = Node.new(c)
      tree[c] = child_node
    end
    if j == 0
      node.child_left = child_node
    else
      tree[l].right_sibling = child_node
    end
    l = c
    tree[c].parent = node
  }
end


for i in 0..(n-1)
  puts tree[i].to_s
end

File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_7_A_RootedTree/ALDS1_7_A.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Tree struct {
	nodes []*Node
}

func NewTree(size int) *Tree {
	nodes := make([]*Node, size)
	return &Tree{nodes}
}

func (t *Tree) append(id int, children []int) {
	var node *Node
	if t.nodes[id] == nil {
		node = &Node{id: id, children: children}
	} else {
		node = t.nodes[id]
		node.children = children
	}
	t.nodes[id] = node
	for _, childID := range children {
		if t.nodes[childID] == nil {
			child := &Node{id: childID}
			t.nodes[childID] = child
		}
		t.nodes[childID].parent = node
	}
}

func (t *Tree) Println() {
	for i := 0; i < len(t.nodes); i++ {
		t.nodes[i].Println()
	}
}

type Node struct {
	id       int
	parent   *Node
	children []int
}

func (n *Node) Println() {
	var d = 0
	p := n.parent
	for p != nil {
		d++
		p = p.parent
	}
	var parentID = -1
	if n.parent != nil {
		parentID = n.parent.id
	}

	cntChildren := len(n.children)

	t := "internal node"
	if parentID == -1 {
		t = "root"
	} else if cntChildren == 0 {
		t = "leaf"
	}

	fmt.Printf("node %d: parent = %d, depth = %d, %s, ", n.id, parentID, d, t)

	fmt.Print("[")
	i := 0
	for i < cntChildren {
		fmt.Printf("%d", n.children[i])
		i++
		if i != cntChildren {
			fmt.Print(", ")
		}
	}
	fmt.Println("]")
}

func nextInt(sc *bufio.Scanner) int {
	sc.Scan()
	n, err := strconv.Atoi(sc.Text())
	if err != nil {
		panic(err)
	}
	return n
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)
	n := nextInt(sc)
	tree := NewTree(n)
	for i := 0; i < n; i++ {
		id := nextInt(sc)
		childrenSize := nextInt(sc)
		children := make([]int, childrenSize)
		for j := 0; j < childrenSize; j++ {
			children[j] = nextInt(sc)
		}
		tree.append(id, children)
	}
	tree.Println()
}
